pkg/rancher-desktop/main/extensions/extensions.ts

```diff
@@ -0,0 +1,207 @@
+            throw ex;
+          }
+        })(),
+
+        // Copy UI
+        (async() => {
+          const uiDir = path.join(workDir, 'ui');
+
+          if (!metadata.ui) {
+            return;
+          }
+
+          await fs.promises.mkdir(uiDir, { recursive: true });
+          await Promise.all(Object.entries(metadata.ui).map(async([name, data]) => {
+            try {
+              await fs.promises.mkdir(path.join(uiDir, name), { recursive: true });
+              await this.client.copyFile(this.id, data.root, path.join(uiDir, name));
+            } catch (ex: any) {
+              throw new ExtensionErrorImpl(ExtensionErrorCode.FILE_NOT_FOUND, `Could not copy UI ${ name }`, ex);
+            }
+          }));
+        })(),
+
+        // Copy host executables
+        (async() => {
+          let plat: 'windows' | 'linux' | 'darwin' = 'windows';
+
+          if (process.platform === 'linux' || process.platform === 'darwin') {
+            plat = process.platform;
+          } else if (process.platform !== 'win32') {
+            throw new Error(`Platform ${ process.platform } is not supported`);
+          }
+          const binDir = path.join(workDir, 'bin');
+
+          await fs.promises.mkdir(binDir, { recursive: true });
+          const binaries = metadata.host?.binaries ?? [];
+          const paths = binaries.flatMap(p => p[plat]).map(b => b?.path).filter(defined);
+
+          await Promise.all(paths.map(async(p) => {
+            try {
+              await this.client.copyFile(this.id, p, binDir);
+            } catch (ex: any) {
+              throw new ExtensionErrorImpl(ExtensionErrorCode.FILE_NOT_FOUND, `Could not copy host binary ${ p }`, ex);
+            }
+          }));
+        })(),
+      ]);
+    } catch (ex) {
+      console.error(`Failed to install extension ${ this.id }, cleaning up:`, ex);
+      await fs.promises.rm(workDir, { recursive: true }).catch((e) => {
+        console.error(`Failed to cleanup extension directory ${ workDir }`, e);
+      });
+      throw ex;
+    }
+
+    await fs.promises.rename(workDir, this.dir);
+    mainEvents.emit('settings-write', { extensions: { [this.id]: true } });
+
+    // TODO: Do something so the extension is recognized by the UI.
+    console.debug(`Install ${ this.id }: install complete.`);
+
+    return true;
+  }
+
+  async uninstall(): Promise<boolean> {
+    const metadata = await this.metadata;
+    const vm = metadata.vm;
+
+    // TODO: Unregister the extension from the UI.
+
+    if ('image' in vm) {
+      console.error('Todo: stop container');
+    } else if ('composefile' in vm) {
+      console.error(`Skipping uninstall of compose file when uninstalling ${ this.id }`);
+    }
+
+    try {
+      await fs.promises.rmdir(this.dir, { recursive: true });
+    } catch (ex: any) {
+      if ((ex as NodeJS.ErrnoException).code !== 'ENOENT') {
+        throw ex;
+      }
+    }
+
+    mainEvents.emit('settings-write', { extensions: { [this.id]: false } });
```

Comment:
NOTE: I'm getting in-editor deprecation warnings for `settings-write`.

Reply:
Hmm, I know this sometimes happens for me until I load `settings.ts` again (to let it update what `Settings` looks like). Does this resolve itself after that?